prov/shm/src/smr_dsa.c

```diff
@@ -267,10 +353,15 @@ static int dsa_idxd_init_wq_array(int shared, int numa_node,
 					"DSA WQ: %s\n",
 					(*dsa_ops.accfg_wq_get_devname)(wq));
 
-			wq_reg = dsa_idxd_wq_mmap(wq);
-			if (wq_reg == NULL)
+			ret = dsa_idxd_wq_mmap(
+				wq, &dsa_context->wq_mmap_support,
+				&dsa_wq_write_support,
+				&dsa_submit_cmd_mode_checked,
+				&mmapped_wq[wq_count], &wq_fd[wq_count]);
+
+			if (ret == -1)
```

Comment:
Starrt with both mmap_suppot = true and wq_write_support = true. Try mapping wq with existing function with condition mmap_support == true. If it returns NULL and current wq_count==0. Set wq_mmap_support false. Write another function to open_wq for write. Try that function with wq_write_support == true. If it returns -1 and wq_count==0, set wq_write_support = false. If first time any of the worked, we use it for subsequent wq. If both fails, wq_count will remain 0.

Reply:
And you don't need dsa_submit_cmd_mode_checked. The second function should check for wq write support. Let's do that for each wq. It is redundant but makes the logic easier to follow.